### PR TITLE
(WIP) [*] CORE : Improved DbQuery class + parameterized queries

### PR DESCRIPTION
**Pull request by firstred**
## Description

Upgrade the `DbQuery` class with some extra functionality. Based on Doctrine's SQL Query Builder.
It is 100% compatible with pre-1.7's `DbQuery` and ~90% compatible with Doctrine's SQL Query Builder.

The `DbQuery` has been designed in such a way that it is very easy to create SQL templates and add parameters in order to execute parameterized queries.

Parameterized queries can lead to a significant performance improvement and is a great defense against SQL injections.

`Db::update` and `Db::insert` are automatically parameterized, thus automatically preventing SQL injections and improving performance with repeated queries.
## How to use it / Best practice

Create an SQL template:

``` PHP
$sql = new DbQuery();
$sql->select('id_product');
$sql->from('product'),
$sql->where('id_product = ? OR id_product = ?');
```

Then set the parameters:

``` PHP
$sql->setParameter(1, Db::TYPE_INTEGER);
$sql->setParameter(2, Db::TYPE_INTEGER);
```

You can also set the parameters as follows:

``` PHP
$sql->setParameters([1,2], [Db::TYPE_INTEGER, Db::TYPE_INTEGER]);
```

The order is important. The available types are limited by `MySQLi`'s types.
More info on (most of) the possibilities can be found here: http://doctrine-orm.readthedocs.org/projects/doctrine-dbal/en/latest/reference/query-builder.html
The compatibility and limitations can be found in the PHPDoc comments.
## Limitations
- PDO's named parameters are not supported, only question mark parameters
- Support for the old `mysql` extension needs to be dropped
## Steps to Test this Feature
1. Try a `SELECT` or `DELETE` through `DbQuery`
2. Try to `UPDATE` or `INSERT` with `Db`

### CODE REVIEW

**Comment by mickaelandrieu** [comment]
Hi @firstred,

can you rebase against develop branch or fix the issue ? 

To me sounds a good idea, but we need time to review this good contribution :)

**Comment by firstred** [comment]
Yeah I can imagine that it takes quite some time. Already took me some time to figure out the best way and on gitter you can see that I basically had already given up on the improvement. :p

**Comment by firstred** [comment]
Has PDO become mandatory? That would make things a lot easier, a LOT! :tada:

**Comment by mickaelandrieu** [comment]
@firstred yeah,  'cause of Doctrine :)

File: classes/db/Db.php [hunk 1]
```diff
@@ -819,7 +944,7 @@ public static function checkEncoding($server, $user, $pwd)
      */
     public static function hasTableWithSamePrefix($server, $user, $pwd, $db, $prefix)
     {
-        return call_user_func_array(array(Db::getClass(), 'hasTableWithSamePrefix'), array($server, $user, $pwd, $db, $prefix));
+        return call_user_func_array([Db::getClass(), 'hasTableWithSamePrefix'], [$server, $user, $pwd, $db, $prefix]);
```
Comment (mickaelandrieu) [inline classes/db/Db.php]: @firstred can you use the old ``array(...)`` notation please ? same for others `[ ]` :)

File: classes/db/DbMySQLi.php [hunk 2]
```diff
@@ -127,6 +163,59 @@ protected function _query($sql)
...
+
+    /**
+     * Execute prepared statement
+     *
+     * @return bool Whether it was executed successfully
+     */
+    protected function _execute()
+    {
+        // Process all params at once before executing
+        $params = [];
```
Comment (mickaelandrieu) [inline classes/db/DbMySQLi.php]: array()

File: classes/db/DbQuery.php [hunk 3]
```diff
@@ -19,81 +19,605 @@
...
+     *
+     * @param string $type SELECT|INSERT|UPDATE|DELETE
      *
      * @return DbQuery
      */
     public function type($type)
     {
-        $types = array('SELECT', 'DELETE');
+        switch ($type) {
+            case 'SELECT':
```
Comment (mickaelandrieu) [inline classes/db/DbQuery.php]: May I suggest: 
```php
$this->type = constant("self::$type");
```

**Comment by mickaelandrieu** [comment]
this is a great job, you may need to rebase in order to relaunch test suite :)

**Comment by firstred** [comment]
Yep, it's going to be a huge update!
Will also try to support named parameters if only PDO is going to be supported.

File: classes/db/Db.php (hunk 1 above)
Comment (firstred) [inline classes/db/Db.php]: Oui! :)

**Comment by kpodemski** [comment]
Is there any way to name parameteres instead of having to bind them one by one in proper order?

I mean if we can do something like:
```
$sql->setParameter('first', Db::TYPE_INTEGER);
$sql->setParameter('second', Db::TYPE_INTEGER);
```

?

**Comment by mickaelandrieu** [comment]
@kpodemski you mean something like ``setParameters()`` ?

**Comment by firstred** [comment]
Both would be nice. E.g. using an associative array for `setParameters()` and `setParameter()` like the way @kpodemski suggested.
The order no longer matters with PDO.

**Comment by firstred** [comment]
Wait... we'll need more than that, we need to pass the name, value and type of the parameter...
Anyone got a good suggestion?

File: classes/db/Db.php [hunk 4]
```diff
@@ -835,7 +960,7 @@ public static function hasTableWithSamePrefix($server, $user, $pwd, $db, $prefix
      */
     public static function checkCreatePrivilege($server, $user, $pwd, $db, $prefix, $engine = null)
     {
-        return call_user_func_array(array(Db::getClass(), 'checkCreatePrivilege'), array($server, $user, $pwd, $db, $prefix, $engine));
+        return call_user_func_array([Db::getClass(), 'checkCreatePrivilege'], [$server, $user, $pwd, $db, $prefix, $engine]);
```
Comment (mickaelandrieu) [inline classes/db/Db.php]: never forget last summer ... I mean last array ^^

File: classes/db/Db.php [hunk 5]
```diff
@@ -848,7 +973,7 @@ public static function checkCreatePrivilege($server, $user, $pwd, $db, $prefix,
      */
     public static function checkAutoIncrement($server, $user, $pwd)
     {
-        return call_user_func_array(array(Db::getClass(), 'checkAutoIncrement'), array($server, $user, $pwd));
+        return call_user_func_array([Db::getClass(), 'checkAutoIncrement'], [$server, $user, $pwd]);
```
Comment (mickaelandrieu) [inline classes/db/Db.php]: same

**Comment by mickaelandrieu** [comment]
@Shudrum I realy want it merged in 1.7 except if you are realy opposed ... I mean this sounds great and I want to complete this work with a DBAL-based implementation.

What do you think ?

**Comment by mickaelandrieu** [comment]
@firstred even Doctrine QueryBuilder dont allow to pass Type in setParameters => https://github.com/doctrine/doctrine2/blob/master/lib/Doctrine/ORM/QueryBuilder.php#L576

**Comment by Shine-neko** [comment]
Big up @firstred @mickaelandrieu  @Shudrum

**Comment by xBorderie** [comment]
@firstred Are you abandonning the idea? Reworking it?

**Comment by firstred** [comment]
Hmmz, I closed it because I think that making a transitional class like this will make us stick to the past.

Why not use Doctrine directly? The sooner we migrate to Symfony components, the better right? 
@Shudrum what do you think?

I've noticed that it's almost impossible to make hybrid 1.5/1.6/1.7 modules anyways, so I'm not sure if this will be useful.

**Comment by molearnik** [comment]
I see this request has been closed and improved DbQuery class won't be merged.
What about @firstred 's suggestion to use Doctrine directly? Could at least AddOns use Doctrine instead of Db class from PrestaShop bundle? 
PrestaShop "Best practices related to Db class" (http://doc.prestashop.com/pages/viewpage.action?pageId=51184692) help to “protect” against SQL injections only by type-casting numerical parameters and by escaping string parameters using pSQL function from "PrestaShop-develop\config\alias.php". That can hardly be considered as perfect defense against SQL injection.
Also, Prestashop's Db class and DbQuery are less convenient than Doctrine.

So, could PrestaShop consider 
1. extending "Best Practices" documentation (on this link:
http://doc.prestashop.com/pages/viewpage.action?pageId=51184692) and officially "approve" Doctrine ORM to AddOn developers?
2. and correspondingly modify AddOns review process in order Doctrine ORM based code was accepted.

**Pull request closed without merging**